proposals/gc/MVP-JS.md

```diff
@@ -0,0 +1,537 @@
+costly wrapper-based JS/WebAssembly interoperability schemes and also gives
+JavaScript a new kind of high-performance object that should be independently
+valuable for optimizing pure JavaScript applications.
+
+The challenge of course is to do this in a way that doesn't inject JS
+peculiarities into the core design of GC in WebAssembly or impact the efficiency
+of WebAssembly GC (even in JavaScript Host Embeddings).
+
+
+## Walkthrough
+
+The JavaScript spec allows [Exotic Objects](https://tc39.github.io/ecma262/#sec-exotic-object)
+to diverge from Ordinary Object behavior by overriding
+[spec-internal methods and slots](https://tc39.github.io/ecma262/#sec-built-in-exotic-object-internal-methods-and-slots).
+An example of this is Typed Arrays, which override various
+[internal property-access methods](https://tc39.github.io/ecma262/#sec-integer-indexed-exotic-objects)
+to turn normal JS property accesses into pure byte-array operations.
+This presently allows JavaScript and WebAssembly to both have high-performance
+access to WebAssembly's [linear memory](https://webassembly.github.io/spec/core/intro/overview.html#memory), without injecting JS
+peculiarities into the design of linear memory. We propose to do likewise for
+WebAssembly structs and arrays.
+
+To do this, we propose to revive, refurbish and repurpose the
+[old Typed Objects proposal](http://smallcultfollowing.com/babysteps/pubs/2014.04.01-TypedObjects.pdf).
+The [new Typed Objects proposal](https://github.com/tschneidereit/proposal-typed-objects/blob/master/explainer.md)
+is designed with WebAssembly in mind and is motivated as follows:
+
+
+### Value Type objects
+
+Corresponding to the set of [field types](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#type-definitions)
+in the WebAssembly GC proposal, the JS Typed Objects proposal includes an
+open-ended set of [Value Type objects](https://github.com/tschneidereit/proposal-typed-objects/blob/master/explainer.md#value-types)
+which contain `[[Read]]` and `[[Write]]` [internal methods](https://tc39.github.io/ecma262/#sec-object-internal-methods-and-internal-slots)
+that convert to and from arbitrary JavaScript values and the pure typed values
+that WebAssembly is defined to interact with.
+
+The JS Typed Object proposal includes a set of [common Value Type objects](https://github.com/tschneidereit/proposal-typed-objects/blob/master/explainer.md#value-types).
+This general set is extended with additional Value Type objects in the
+`WebAssembly` namespace [described below](#other-webassembly-value-types), so
+that all WebAssembly field types have a corresponding Value Type object. The 
+generally-applicable Value Type objects defined by the JS Typed Object proposal
+are broken into two categories:
+
+* *Primitive* Value Types, exposed as singleton objects: `uint8`,
+  `uint16`, `uint32`, `uint64`, `int8`, `int16`, `int32`, `int64`, `float32`,
+  `float64`, `any`, `string`.
```

Comment:
There is indeed `string` Value Type in [Typed Objects proposal](https://github.com/tschneidereit/proposal-typed-objects/blob/master/explainer.md#value-types) and there is also `object` Value Type which is missing from Value Types list here. Both `string` and `object` are confusing.

Reply:
The idea is that these Value Type objects are defined as part of the general "Typed Objects" JS feature and so the meaning of these value types is entirely JS-specific.  To access objects with these field types, wasm would have to import these types.  Static wasm type checking + dynamic coercions at the wasm/JS boundary would ensure that `string` values were always, in fact, JS strings, and `object` values were always, in fact, JS objects (i.e., `typeof(v) === "object" && v !== null`).  At instantiation time, this could eliminate what would otherwise be the runtime checks required when passing `anyref` to methods taking strings/objects.

Does that make sense?

Reply:
It does.

`object` value is still missing from list of Value Types here though.

Reply:
Oops, that was an unintentional omission.  Fixed and thanks!

Reply:
Resolving, but feel free to reopen if this wasn't answered.